tests/test_lzip_decoder.c

```diff
@@ -0,0 +1,454 @@
+///////////////////////////////////////////////////////////////////////////////
+//
+/// \file       test_lzip_decoder.c
+/// \brief      Tests decoding lzip data
+//
+//  Author:     Jia Tan
+//
+//  This file has been put into the public domain.
+//  You can do whatever you want with this file.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include "tests.h"
+
+#ifdef HAVE_LZIP_DECODER
+
+// Memlimit large enough to pass all of the test files
+#define MEMLIMIT 1 << 20
+#define DECODE_CHUNK_SIZE 1024
+
+
+// Helper function to decode a good file with no flags and plenty
+// high memlimit
+static void
+basic_lzip_decode(const char* src, const uint8_t* expected_result,
+		size_t expected_result_size) {
+	size_t file_size;
+	uint8_t *data = tuktest_file_from_srcdir(src, &file_size);
+
+	lzma_stream strm = LZMA_STREAM_INIT;
+	assert_lzma_ret(lzma_lzip_decoder(&strm, MEMLIMIT, 0), LZMA_OK);
+
+	uint8_t *output_buffer = tuktest_malloc(expected_result_size);
+
+	strm.next_in = data;
+	strm.next_out = output_buffer;
+	strm.avail_out = expected_result_size;
+
+	// Feed 1 byte at a time to the decoder to look for any bugs
+	// when switching between decoding sequences
+	lzma_ret ret = LZMA_OK;
+	while (ret == LZMA_OK) {
+		strm.avail_in = 1;
+		ret = lzma_code(&strm, LZMA_RUN);
+	}
+
+	assert_lzma_ret(ret, LZMA_STREAM_END);
+
+	assert_uint_eq(strm.total_in, file_size);
+	assert_uint_eq(strm.total_out, expected_result_size);
+	assert_array_eq(output_buffer, expected_result,
+			expected_result_size);
+	lzma_end(&strm);
+}
+
+
+static void
+test_options(void)
+{
+	// Test NULL stream
+	assert_lzma_ret(lzma_lzip_decoder(NULL, MEMLIMIT, 0),
+			LZMA_PROG_ERROR);
+
+	// Test invalid flags
+	lzma_stream strm = LZMA_STREAM_INIT;
+	assert_lzma_ret(lzma_lzip_decoder(&strm, MEMLIMIT, UINT32_MAX),
+			LZMA_OPTIONS_ERROR);
+	// Memlimit tests are done elsewhere
+}
+
+
+static void
+test_v0_decode(void) {
+	// This tests if liblzma can decode lzip version 0 files.
+	// lzip 1.17 and older can decomperss this, but lzip 1.18
+	// and newer can no longer decode these files.
+	const char expected_result[] = "Hello\nWorld!\n";
+	basic_lzip_decode("files/good-1-v1.lz",
```

Comment:
The test files are tested by test_files.sh already. Replacing test_files.sh by something else is one question but duplicating the test here doesn't seem useful.

Reply:
Its useful if we break something in the lzip decoder that causes this to fail. This will tell us the lzma_ret value, but the test_files.sh won't. Also, this checks the actual contents of the file match, but test_files.sh just check if the decoder returns success or not.

Reply:
I made a typo anyway. It should be:
	basic_lzip_decode("files/good-1-v0.lz",
                        (const uint8_t *) expected_result,
			sizeof(expected_result) - 1);

Reply:
Content checking is done by test_files.sh thanks to CRC32, that part is good enough. The return value isn't checked by test_files.sh, that is true. I'm not sure how important it is but it's true it's not much code to check it here to be sure.

Reply:
Maybe this is too philosophical about the purpose of the tests, but using the checksum as the content verification assumes that the checksum was properly checked. The more paranoid approach is to also verify the checksum value

Reply:
I don't disagree, it's just that I feel it's not that important for these specific tests to be extremely paranoid. Using lzma_crc32() here after decompression to check the uncompressed data might work too? I mostly would like to avoid the need to have the full uncompressed file content in the C code but since that is fairly short I suppose it's OK too, it just has to be in hex so that it survives conversion to EBCDIC. So if you prefer that it's OK.